Verdi node delete 923

lekah opened the pull request:
verdi node delete via command line is introduced, making the secret script not longer necessary.
Made a test for that functionality and updated the documentation.

nmounet commented on aiida/cmdline/commands/node.py:
```diff
@@ -303,6 +306,60 @@ def print_node_info(self, node, print_groups=False):
+            action='store_true')
+        # Commenting this option for now
+        # parser.add_argument('-f', '--force', help='force deletion, disables final user confirmation', action='store_true')
+        parser.add_argument('-v', '--verbosity', help='verbosity level', action='count', default=1)
```
for the verbosity level: not sure it's very clear like this...

lekah replied:
This has been addressed in 3fcab81

szoupanos commented on aiida/backends/sqlalchemy/utils.py:
```diff
@@ -533,3 +533,20 @@ def alembic_command(selected_command, *args, **kwargs):
+
+    session = sa.get_scoped_session()
+
+    with session.begin(subtransactions=True):
```
At aiida.orm.implementation.sqlalchemy.code.delete_code I see the following 

```
        from aiida.backends.sqlalchemy import get_scoped_session
        session = get_scoped_session()
        session.begin(subtransactions=True)
        try:
            code.dbnode.delete()
            session.commit()
        except:
            session.rollback()
            raise
```

Which should be the verbose way of what you wrote. I suppose that at the end of the with clause/statement, there is an auto-commit (according to http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#session-autocommit).

It is worth verifying that for any kind of error in the queries, there is a rollback.I suppose there is but...
The same for Django

lekah replied:
The django part is taken from the existing script, and the same logic is replicated for sqlalchemy.
The django code is working on big databases, I was using it in a while in the current form. 
Regarding your comment on verbosity, I disagree: I do explicit deletion of the links, since it's more robust than relying on deletion cascades (which I'm not sure is implemented) for the links. Maybe above example for code deletion should be checked?

From the docs that you cite I understand that using the session in a try-except clause or with a context-manager is equivalent. As for the 'it is worth verifying': That's the reviewers job ;)

szoupanos replied:
It is not actually a proof of correctness that it runs on big databases (it's an indication). I also never mentioned to do deletion propagation etc.

An overall comment. The reviewers' job is to make some constructive comments on the code and highlight some potential corner cases (of course they should be somehow answered constructively since the author asked the reviewer's opinion & the reviewer invested some time to give feedback). 

Unfortunately, I don't have the time to check your code for all these corner cases.

lekah replied:
Me see problem, I make fix... Seriously, let's not have a big discussion here, about proofs of correctness, and the role of reviewers. I added  the explicit try-and-except clause that you wanted and made some more tests with deliberately wrong queries. The rollback works. Actually, I found that, different from Django, the cascading is not implemented for group-membership, so I also delete on the dbnode_dbgroup table  in 83b0478 .

szoupanos replied:
Thanks for the changes

lekah commented:
Welcome to the party @giovannipizzi 
I tried to make the command both transparent in usage and safe.
By default, it asks for user confirmation before deleting anything. Also, there is a dry-run option.
This particular feature has been requested by a multitude of users. I agree that it's dangerous, which is why the precautions and prompt for user confirmation. 
I'm very open to putting another confirmation question, and/or printing a warning in all-caps.
(ARE YOU REALLY SURE YOU WANT TO DELETE X NODES?)

giovannipizzi commented:
I gave my comments to @lekah personally, after which he did the two new commits. I think the way it is now it's acceptable. Note that the follow-returns has been removed from the command line (I think it's too dangerous, it's not what people want). It's still in the code to avoid to remove the implementation)>

giovannipizzi commented:
Important question: what to do when deleting outputs of sealed calculations? @sphuber @muhrin 
It shouldn't be allowed, what do you think? Otherwise the calculation looks ok but it misses some important output. Is this a constraint we want to have? As a consequence, should there be a flag for deleting 'creating' calculations?

giovannipizzi commented:
After further discussion with @lekah we realised that there might be a number of reasons why users want to delete outputs. They should at least be aware that they are doing it. So, proposed solution:

1. there is a `check` flag, True by default, that checks 
   - if there are any Data node to be deleted, CREATEd by a calculation that is *not* going to be deleted
   - if there are any Calculation node to be deleted, CALLed by a calculation that is *not* going to be deleted
2. In this case, it will print out a message with the pairs of nodes (input -> output) and a message explaining why in general this is discouraged
3. If the user, anyway, wants to continue, we add an entry to the CalculationLog table (visible with `verdi calculation logshow` or `verdi work report`) that says that the user asked to delete an output, similarly to what is done when a calculation is killed - so at least it is written down, and when checking back old calculations, if something weird is seen, i.e. missing outputs, one can check why this is so

sphuber commented:
Do these final checks hold for all calculations or only sealed ones? Do we now have a clear definition and implementation of the `sealed` concept. I for one am not sure what exactly it entails.

lekah commented:
For all calculations

lekah commented:
Commit c0da5c4c addresses the first set of remarks by @giovannipizzi . There is a query to find all calculations that lose created data, and a query for calculations that lose called instances.
prompts to stdout inform the user, and in case of deletion, information is written to DbLog.
No check-flag yet, this will follow asap.

lekah commented:
Current link types are enumerated (literally) here: http://aiida-core.readthedocs.io/en/latest/_modules/aiida/common/links.html
For how these are created, check for example the Node._add_dblink_from method and everything that calls it, if you're interested in higher-level implementation.  @sphuber , is there a draft on currently existing link types?

sphuber commented:
If with 'draft' you are referencing to the document that I started writing and presented at some point, that is all still purely hypothetical and none of it has been implemented. You already pointed to the link types that currently exist in `v0.11.0`. As to how they are used, the best thing would be to look at the discussion in issue #687 where we defined the migration rules to retroactively write the link types for older databases. These should reflect the rules that the code currently uses to create new links.

lekah commented:
71cb8dfe addresses the comments and suggestions by @giovannipizzi . Calculations will have it written to the log if created or called instances are deleted. There is a flag to disable this checks, but it cannot be enabled via command-line.

giovannipizzi commented:
I don't merge in case someone wants to give some final comment.

szoupanos commented:
Yes, I agree. I also had a short chat with Leo and I also agree with what @giovannipizzi said.
We can add more checks (if we discover that we may want them - something is missing) in the future,
